jsoniter-scala-examples/src/main/scala/com/github/plokhotnyuk/jsoniter_scala/examples/Example01.scala

```diff
@@ -1,23 +1,74 @@
 package com.github.plokhotnyuk.jsoniter_scala.examples
 
+import java.nio.charset.StandardCharsets
+
 import com.github.plokhotnyuk.jsoniter_scala.macros._
 import com.github.plokhotnyuk.jsoniter_scala.core._
 
 /**
   * Example of basic usage from README.md
   */
 object Example01 {
-  case class Device(id: Int, model: String)
+  sealed trait NameEnum
+  final object TEMP extends NameEnum
+  final object ID extends NameEnum
+
+  sealed trait StateEnum
+  final object STOP extends StateEnum
+  final object NIGHT extends StateEnum
+
+  sealed trait Leaf
+  final case class IntLeaf(value: Int) extends Leaf
+  final case class StringLeaf(value: String) extends Leaf
+  final case class FloatLeaf(value: Float) extends Leaf
 
-  case class User(name: String, devices: Seq[Device])
+  final case class Elmnt(value: Leaf,
+                         timestamp: Long,
+                         name: NameEnum,
+                         state: StateEnum,
+                         unit: Option[String])
+
+  implicit val leafCodec: JsonValueCodec[Leaf] = new JsonValueCodec[Leaf] {
+      def nullValue: Leaf = null
+      def decodeValue(in: JsonReader, default: Leaf): Leaf = in.nextToken() match {
+        case '"' =>
+          in.rollbackToken()
+          StringLeaf(in.readString(null))
+        case _ =>
+          in.rollbackToken()
+          val d = in.readDouble() // use in.readBigDecimal(null) for exact precision
+          val i = d.toInt
+          if (d == i) IntLeaf(i)
```

Comment:
If we have `"value": 0.0` it will parse it as `IntLeaf`. Just found. I will try to fix it unless you have a better idea.

Reply:
Please check the following commit if it works for you: https://github.com/plokhotnyuk/jsoniter-scala/pull/509/commits/d56102d174d9eb8b8b1fb9ab06582834327cc69c

Now expected output is:
```
List(Elmnt(FloatLeaf(0.0),1587216049012,com.github.plokhotnyuk.jsoniter_scala.examples.Example01$TEMP$@2c8d66b2,com.github.plokhotnyuk.jsoniter_scala.examples.Example01$NIGHT$@5a39699c,None), Elmnt(StringLeaf(4105f527-69dc-4a4c-ab84-918b256c7dc0),1587216049003,com.github.plokhotnyuk.jsoniter_scala.examples.Example01$ID$@3cb5cdba,com.github.plokhotnyuk.jsoniter_scala.examples.Example01$STOP$@56cbfb61,None))
[{"value":0.0,"timestamp":1587216049012,"name":"TEMP","state":"NIGHT"},{"value":"4105f527-69dc-4a4c-ab84-918b256c7dc0","timestamp":1587216049003,"name":"ID","state":"STOP"}]
```